gossip/tx_scrambler_test.go

```diff
@@ -0,0 +1,597 @@
+		{hash: common.Hash{2}},
+		{hash: common.Hash{1}},
+	}
+
+	shuffleEntries(entries)
+	result, _, _ := analyseEntryList(entries)
+	if len(result) == 0 {
+		t.Fatal("analyseEntryList returned empty list")
+	}
+
+	seen := map[common.Hash]struct{}{}
+	for _, entry := range result {
+		if _, seen := seen[entry.hash]; seen {
+			t.Fatalf("duplicate hash %v", entry.hash)
+		}
+		seen[entry.hash] = struct{}{}
+	}
+}
+
+func TestTxScrambler_UnifyEntries_SaltCreationIsDeterministic(t *testing.T) {
+	entries := []*scramblerEntry{
+		{hash: common.Hash{1}},
+		{hash: common.Hash{2}},
+		{hash: common.Hash{3}},
+		{hash: common.Hash{2}},
+		{hash: common.Hash{1}},
+	}
+
+	_, wantedSalt, _ := analyseEntryList(entries)
+	for range 10 {
+		shuffleEntries(entries)
+		_, gotSalt, _ := analyseEntryList(entries)
+		if gotSalt != wantedSalt {
+			t.Fatal("incorrect salt - salt creation is not deterministic")
+		}
+	}
+
+}
+
+func TestTxScrambler_AnalyseEntryList_ReportsDuplicateAddresses(t *testing.T) {
+	tests := []struct {
+		name         string
+		input        []*scramblerEntry
+		hasDuplicate bool
+	}{
+		{
+			name: "has duplicate address",
+			input: []*scramblerEntry{
+				{sender: common.Address{1}},
+				{sender: common.Address{3}},
+				{sender: common.Address{2}},
+				{sender: common.Address{3}},
+			},
+			hasDuplicate: true,
+		},
+		{
+			name: "has no duplicate address",
+			input: []*scramblerEntry{
+				{sender: common.Address{1}},
+				{sender: common.Address{2}},
+				{sender: common.Address{3}},
+			},
+			hasDuplicate: false,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			_, _, hasDuplicateAddresses := analyseEntryList(test.input)
+			if hasDuplicateAddresses != hasDuplicateAddresses {
+				t.Error("wrongly reported duplicate address")
+			}
+		})
+	}
+
+}
+
+func TestTxScrambler_ScrambleTransactions_ScrambleIsDeterministic(t *testing.T) {
+	res1 := []*scramblerEntry{
+		{hash: common.Hash{1}},
+		{hash: common.Hash{2}},
+		{hash: common.Hash{3}},
+		{hash: common.Hash{2}},
+		{hash: common.Hash{1}},
+	}
+
+	res2 := deepCopyEntries(res1)
+
+	for i := 0; i < 10; i++ {
+		salt := createRandomSalt()
+		scrambleTransactions(res1, salt)
+		for j := 0; j < 10; j++ {
+			shuffleEntries(res2)
+			scrambleTransactions(res2, salt)
+			if !reflect.DeepEqual(res1, res2) {
+				t.Error("scramble is not deterministic")
+			}
+		}
+	}
+}
+
+func TestTxScrambler_SortTransactionsWithSameSender_SortsByNonce(t *testing.T) {
+	entries := []*scramblerEntry{
+		{
+			hash:   common.Hash{1},
+			sender: common.Address{1},
+			nonce:  1,
+		},
+		{
+			hash:   common.Hash{2},
+			sender: common.Address{2},
+			nonce:  1,
+		},
+		{
+			hash:   common.Hash{3},
+			sender: common.Address{3},
+			nonce:  1,
+		},
+		{
+			hash:   common.Hash{4},
+			sender: common.Address{2},
+			nonce:  2,
+		},
+		{
+			hash:   common.Hash{5},
+			sender: common.Address{1},
+			nonce:  2,
+		},
+	}
+
+	sortTransactionsWithSameSender(entries)
```

Comment:
The input is already sorted in the desired way. So if this function does nothing, the test passes.

Reply:
Added shuffle before calling `sortTransactionsWithSameSender`